src/main/java/org/truffleruby/core/string/StringNodes.java

```diff
@@ -2162,6 +2162,42 @@ private int codePointX(Encoding enc, CodeRange codeRange, byte[] bytes, int p, i
         }
     }
 
+    @CoreMethod(names = "undump", taintFrom = 0)
+    @ImportStatic(StringGuards.class)
+    public abstract static class UndumpNode extends CoreMethodArrayArgumentsNode {
+        @Specialization(guards = "isAsciiCompatible(string)")
+        protected RubyString undumpAsciiCompatible(RubyString string,
+                @CachedLanguage RubyLanguage language,
+                @Cached AllocateHelperNode allocateHelperNode,
+                @Cached RopeNodes.MakeLeafRopeNode makeLeafRopeNode) {
+            // Taken from org.jruby.RubyString#undump
+            RopeBuilder outputBytes = StringSupport.undump(string.rope, getContext(), this);
+
+            final Rope rope = makeLeafRopeNode
+                    .executeMake(
+                            outputBytes.getBytes(),
+                            outputBytes.getEncoding(),
+                            CR_UNKNOWN,
+                            outputBytes.getLength());
+
+            final RubyClass logicalClass = coreLibrary().stringClass;
+            final Shape shape = allocateHelperNode.getCachedShape(logicalClass);
+            final RubyString result = new RubyString(logicalClass, shape, false, false, rope);
+            allocateHelperNode.trace(result, this, language);
```

Comment:
Could you use `StringOperations.createString(RubyLanguage, RubyContext, AllocateHelperNode, RubyContextSourceNode, Rope)` instead? (since the return value class is always `String`)

Reply:
Seems like both this and the rope length issue are resolved by simply using `MakeStringNode#fromBuilder` (unless I'm wrong about what it does).

Reply:
Indeed, that's even better :)